### PR TITLE
New plots for convergence test using Plots.jl

### PR DESCRIPTION
**Pull request by francispoulin**
This has been tested using the advection schemes but should be easy enough to generalize to other tests.

### CODE REVIEW

**Comment by francispoulin** [comment]
The plots for the different advection schemes are copied below.

**CenteredSecondOrder**
convergence_ratesCenteredSecondOrder

**CenteredFourthOrder**
convergence_ratesCenteredFourthOrder

**UpwindBiasedThirdOrder**
convergence_ratesUpwindBiasedThirdOrder

**UpwindBiasedFifthOrder**
convergence_ratesUpwindBiasedFifthOrder

**WENO5
!**[convergence_ratesWENO5](https://user-images.githubusercontent.com/8239041/102395470-2bb16100-3fa9-11eb-85b0-5cd9b74bf22b.png)

**Comment by francispoulin** [comment]
If I wanted to check the formula for UpwindBiasedFifthOrder, to confirm that everything looks good, can someone point me to a reference in the literature?

**Comment by glwagner** [comment]
I think you might be able to derive the fifth order scheme from Wicker and Skamarock (2002):

https://palm.muk.uni-hannover.de/trac/wiki/doc/tec/discret

http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.119.8289&rep=rep1&type=pdf

But I'm not 100% sure...

**Comment by francispoulin** [comment]
> I think you might be able to derive the fifth order scheme from Wicker and Skamarock (2002):
> 
> https://palm.muk.uni-hannover.de/trac/wiki/doc/tec/discret
> 
> http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.119.8289&rep=rep1&type=pdf
> 
> But I'm not 100% sure...

Very helpful.  In the left base interpolant, I am pretty sure the second last coefficient should be -3 and not -13.  Probably a similar change for the right based interpolant.

**Comment by francispoulin** [comment]
I decided to program the different advection schemes in Julia, without using Oceananigans, and I found that the 5th order upwinding stencil, as used in Oceananigans, should have fifth order convergence.  See below.

```
For UpwindingFirstOrder: Rate of convergence = -0.9661593569685473 expected = 1

For CenterSecondOrder: Rate of convergence = -1.9640228832659088 expected = 1

For UpwindingThirdOrder: Rate of convergence = -2.9708218645270197 expected = 3

For CenterFourthOrder: Rate of convergence = -3.7210792193087565 expected = 4

For UpwindingFifthOrder: Rate of convergence = -4.847601153924809 expected = 5

For CenterSixthOrder: Rate of convergence = -5.784776184328756 expected = 6

```

**Comment by glwagner** [comment]
Heh I think if you can produce a convergence plot then its worthy to be included in the code! Someone might want 6th order advection...

> The bad news is that I don't know why we are not getting the correct values in Oceananigans, even though I am using the exact same tests.

That sounds like you've isolated a bug...

**Comment by glwagner** [comment]
> I admit that I don't understand the symmetric interpolant stuff that is done, which is perhaps why I can't see the error.

This distinction only matters for momentum advection. For tracer advection, the velocity field does not have to be interpolated to the location of the tracer fluxes, since it's already located there due to the staggered arrangement of velocity fields and tracer fields.

For tracer advection, all that matters are the definition of the left-biased and right-biased functions for interpolating cell-averaged tracer values to cell faces. These are the functions `left_biased_interpolate_xᶠᵃᵃ`, etc.

**Comment by francispoulin** [comment]
I have written a script, with much help from @ali-ramadhan  and @glwagner , that computes the rates of convergence for the one dimensinal constant advection case using two approaches: the 1st to 6th order schemes that I mentioned before, and the 2nd to 5th order schemes that are in Oceananigans.  The results are shown below.

convergence_rates_all

convergence_rates_Oceananigans

**Good news:** Oceananigams produces the correct slopes within error for all the cases.

**Next problem:** Figure out why we didn't get the right slope in the other calculation from before and fix whatever the bug might be.

You can find the code [here](https://github.com/CliMA/Oceananigans.jl/blob/fjp/update-convergence-tests/validation/new_framework/rates_of_convergence.jl)

**Comment by francispoulin** [comment]
Below you will see two plots of the rates of convergence rusing schemes that exist in Oceananigans.  

This is using schemes that I programmed up myself that I ensured used double precision and they give the expected slopes, as they should.

convergence_rates

This similar plot but using the advection schemes in Oceananigans.  The slopes in the legend are taken from the above and are not the slopes of the curves from Oceananigans.

convergence_rates_oceananigans

You will observe that they agree very well at coarser resolutions but Oceananigans flattens out at a certain level.  I think it's because somewhere in the code we are using single instead of double, and hence why it can't do any better.

Does anyone have any ideas where we might be lowering the accuracy from double to single?

In case you are interested, these are the calculations of the rates of convergence for the two sets of calculations.

```
Method = Center2ⁿᵈ,           Rate of Convergence = 1.99, Expected = 2
Method = CenteredSecondOrder(), Rate of Convergence = 1.99, Expected = 2
Method = Upwind3ʳᵈ,           Rate of Convergence = 2.99, Expected = 3
Method = UpwindBiasedThirdOrder(), Rate of Convergence = 3.20, Expected = 3
Method = Center4ᵗʰ,           Rate of Convergence = 3.98, Expected = 4
Method = CenteredFourthOrder(), Rate of Convergence = 2.45, Expected = 4
Method = Upwind5ᵗʰ,           Rate of Convergence = 4.97, Expected = 5
Method = UpwindBiasedFifthOrder(), Rate of Convergence = 1.67, Expected = 5
```

**Comment by francispoulin** [comment]
I would like to suggest the following checklist moving forward

- [ ] Decide whether we like the structure of the script `new_rates_of_convergence.jl`.  How can we might improve it?
- [ ] Find where the bottleneck is in Oceanangans to make things globally single precision.
- [ ] Adapt this approach to the other validation cases that currently exist.
- [ ] Make `UpwindFirstOrderBiased()` advection scheme.
- [ ] Make `CenteredSixthOrder()` advection scheme.

**Review by ali-ramadhan — COMMENTED** [review]
Definitely agree with @glwagner that the more convergence tests, the better!

I think this PR might end up conflicting with PR https://github.com/CliMA/Oceananigans.jl/pull/1223 which refactors the convergence tests to get them working again and adds a Buildkite pipeline for validation experiments (completely my fault leaving it open for so long).

Happy to work together to make sure everything we develop in the two PRs ends up on the Buildkite pipeline!

> Decide whether we like the structure of the script `new_rates_of_convergence.jl`. How can we might improve it?

I think it looks pretty good! Skimming through it, it seems to me like it's replacing https://github.com/CliMA/Oceananigans.jl/blob/master/validation/convergence_tests/one_dimensional_advection_schemes.jl? This would be a good thing since that script is currently broken I think (and it's not touched in PR #1223).

I think it would be good to add some `@test`s checking various metrics, e.g. the rates of convergence are as expected, so that we can run these convergence tests as part of a Buildkite pipeline (convergence tests tend to go stale and break if we don't test them).

> Bad news: Some part of Oceananigans (maybe times-stepping?) must use single precision accuracy, and that truncates the error of the method as a whole.
> 
> Question: where is the bottleneck that reduces the global spatial accuracy from double to single precision?

I'm not aware of any single-precision bottleneck, if anything there are probably places in the code that assume `Float64` even if you try to set up a `Float32` model.

I do agree that other errors might be starting to dominate at high resolution. When first performing the 1D advection scheme convergence tests, we found that the rates of convergence for higher-order advection schemes plateaued at high resolution. We later found out it was because of the AB2 time-stepping which is actually 1st-order accurate, and using RK3 time-stepping helped a lot. Also reducing the integration time (i.e. integrating for like 10 time steps instead of 1000+) and advecting a cosine instead of a Gaussian helped too.

Probably a stupid question but I expected straight lines for the "expected rates of convergence" since it's a log-log plot (even if it's base 10 one one axis and base 2 on another) but they seem slightly curved?

**Comment by francispoulin** [comment]
Thanks @ali-ramadhan for the feedback.

What I have so far is only redoing the `one_dimensional_advection_schemes.jl` but I thought this was a good place to start.  When we are happy with this it should be easy to do other more interesting cases.

I am happy to work with you to merge this with what you did in the other PR, when you have time.

I agree that having some `@test` checks is a good idea. 

I don't think it's as simple as everything being single as I tried running Oceananigans using single and reproduced the same plot.  @glwagner suggested maybe something happens in the pressure solve.  Could be but I don't know yet but I'm going to try and test this using ShallowWaterModel, which  has no pressure solve.

I have used a small enough time step that the scheme hasn't bee a problem, but I agree that RK3 would be better.  I was only  using QAB2 because I wanted to compare it with what I had programmed, AB2, to get similar results.  From what I have seen both QAB2 and AB2 do pretty much the same thing, in this example.

**Comment by francispoulin** [comment]
When I use the advection schemes from ShallowWaterModel I get that the solution computed from Oceananigans has the correct slopes going down to very fine accuracies.  The two plots are identicial.

I think this is further evidence that the lack of accuracy comes from the pressure solve.

One possibility is to have these convergence tests done through ShallowWaterModel, at least for the mean time.

File: src/Models/ShallowWaterModels/solution_and_tracer_tendencies.jl [hunk 1]
```diff
@@ -62,8 +62,24 @@ Compute the tendency for the height, h.
...
 
+    U = 1
+
+    flux = zeros(grid.Nx+1, grid.Ny, grid.Nz)
+    
+    for i in 1:grid.Nx+1, j in 1:grid.Ny, k in 1:grid.Nz
+        
+        flux[i, j, k] = advective_tracer_flux_x(i, j, k, grid, advection, U, solution.h)
+        
+    end
```
Comment (glwagner) [inline src/Models/ShallowWaterModels/solution_and_tracer_tendencies.jl:73]: @francispoulin you don't want this loop in the tendency function. I think you just want

```julia
return advective_tracer_flux_x(i, j, k, grid, advection, U, solution.h)
```

Comment (francispoulin) [reply src/Models/ShallowWaterModels/solution_and_tracer_tendencies.jl:73]: Agreed.  I will make that change today and update the PR

**Comment by francispoulin** [comment]
I installed Julia and set up Oceananigans on my laptop and when I try running `test_shallow_water_mode.jl` I get an error.  I suspect this is because of my laptop more than the PR but any ideas what's going on here?  

```[2021/01/07 11:52:45.500] INFO    Testing time-stepping ShallowWaterModels [CPU(), (Periodic, Periodic, Bounded)]...
Time-stepping ShallowWaterModels [CPU(), (Periodic, Periodic, Bounded)]: Error During Test at /home/fpoulin/software/Oceananigans.jl/test/test_shallow_water_models.jl:82
  Test threw exception
  Expression: time_stepping_shallow_water_model_works(arch, topo, nothing)
  TaskFailedException:
  BoundsError
  Stacktrace:
   [1] getindex at ./number.jl:83 [inlined]
   [2] advective_tracer_flux_x at /home/fpoulin/software/Oceananigans.jl/src/Advection/upwind_biased_advective_fluxes.jl:105 [inlined]
   [3] h_solution_tendency at /home/fpoulin/software/Oceananigans.jl/src/Models/ShallowWaterModels/solution_and_tracer_tendencies.jl:71 [inlined]
   [4] cpu_calculate_Gh! at /home/fpoulin/.julia/packages/KernelAbstractions/jAutM/src/macros.jl:230 [inlined]
   [5] __thread_run(::Int64, ::Int64, ::Int64, ::KernelAbstractions.Kernel{KernelAbstractions.CPU,KernelAbstractions.NDIterat```

**Comment by glwagner** [comment]
I think the problem might be the loop that you are executing inside a kernel.

**Comment by francispoulin** [comment]
A question for @glwagner .

At the moment the tendencies in the evolution equation for height is computed here using essentailly a second-order finite difference method.  This is fine for a start, but as discussed elsehwere, it would be better to use finite volume and therefore using the advection schemes that already exist.

https://github.com/CliMA/Oceananigans.jl/blob/fjp/update-convergence-tests/src/Models/ShallowWaterModels/solution_and_tracer_tendencies.jl#L71

In line 69 above you will see that I tried doing this by definig a new operator, div_UV, for partial_x U + partial_y V.  This was clunky because I needed to pass variables U and V, and I had an error.  The details can be found here,

https://github.com/CliMA/Oceananigans.jl/blob/9402a90af072a495ef130d540661be9327a51552/src/Models/ShallowWaterModels/shallow_water_advection_operators.jl#L24

Can you suggest a better way of doing this?

**Comment by francispoulin** [comment]
This will inspire a new PR to show some of these results.

**Pull request closed without merging**